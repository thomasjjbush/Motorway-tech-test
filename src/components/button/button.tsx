import React, { FC, ReactElement } from 'react';
import styled from 'styled-components';
import { ButtonProps as Props, StyledProps } from '../../../types';

const StyledButton = styled.button<StyledProps>`
    appearance: none;
    background-color: ${({ theme }) => theme.colors.blue};
    border: 0;
    border-radius: 5px;
    color: white;
    cursor: pointer;
    font-size: 16px;
    padding: 10px 20px;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
`;

export const Button: FC<Props> = ({ children, disabled = false, onClick, type = 'button' }: Props): ReactElement => (
    <StyledButton disabled={disabled} onClick={onClick} type={type}>
        {children}
    </StyledButton>
);
