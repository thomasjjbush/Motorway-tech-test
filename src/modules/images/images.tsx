import React, { FC, ReactElement, useEffect, useState } from 'react';
import styled from 'styled-components';
import { server } from '../../utils/server/server';
import { noOfImages, serverDomain } from '../../const/const';
import { Button } from '../../components/button/button';
import { Image } from '../../components/image/image';
import { Image as ImageT, ImagesProps as Props, StyledProps } from '../../../types';

const Grid = styled.div<Props>`
    display: grid;
    grid-gap: 20px;
    grid-template-columns: repeat(${({ columns }): number => columns}, 1fr);
    margin: auto;
    max-width: 1300px;
`;

const User = styled.div<StyledProps>`
    align-items: center;
    background-color: ${({ theme }) => theme.colors.yellow};
    bottom: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    left: 0;
    opacity: 0;
    padding: 20px;
    position: absolute;
    right: 0;
    transition: opacity 0.5s ease;
`;

const GridItem = styled.div`
    margin-bottom: 20px;
    position: relative;

    :hover,
    :focus {
        img {
            filter: grayscale(100%);
        }

        > ${User} {
            opacity: 1;
        }
    }
`;

const Name = styled.h2`
    margin: 0;
`;

const Likes = styled.p`
    margin: 0;
`;

export const Images: FC<Props> = ({ columns }: Props): ReactElement => {
    const [images, setImages] = useState<ImageT[]>([]);
    const [offset, setOffset] = useState(0);

    useEffect(() => {
        (async () => {
            setImages(await server('images', noOfImages, 0));
        })();
    }, []);

    const onClick = async () => {
        setOffset(offset + noOfImages);
        setImages([...images, ...(await server('images', noOfImages, offset + noOfImages))]);
    };

    return (
        <Grid columns={columns}>
            {images
                .reduce(
                    (acc, img, i) => acc[i % columns].push(img) && acc,
                    [...new Array(columns)].map(() => []),
                )
                .map((column: ImageT[], i) => {
                    return (
                        <div key={i}>
                            {column.map(({ alt_description, id, likes, url: endpoint, user }) => (
                                <GridItem data-test-id="grid-item" key={id} tabIndex={0}>
                                    <Image alt={alt_description} src={`${serverDomain}${endpoint}`} />
                                    <User>
                                        <Image
                                            alt={user.name}
                                            round
                                            src={`${serverDomain}${user.profile_image}`}
                                            width="75px"
                                        />
                                        <Name>{user.name}</Name>
                                        <Likes>Likes: {likes}</Likes>
                                    </User>
                                </GridItem>
                            ))}
                        </div>
                    );
                })}
            <Button onClick={onClick} type="button">
                Load more
            </Button>
        </Grid>
    );
};
