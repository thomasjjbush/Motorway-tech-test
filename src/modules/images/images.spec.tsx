import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Images } from './images';
import { Button } from '../../components/button/button';

const images = [
    {
        alt_description: 'alt description',
        id: '1',
        likes: 500,
        url: '/endpoint',
        user: {
            name: 'Lord Souron',
            profile_image: '/profile_image_souron',
        },
    },
    {
        alt_description: 'another alt description',
        id: '2',
        likes: 500,
        url: '/another-endpoint',
        user: {
            name: 'Samwise the brave',
            profile_image: '/profile_image_sam',
        },
    },
    {
        alt_description: 'yet another alt description',
        id: '3',
        likes: 500,
        url: '/uyet-another-endpoint',
        user: {
            name: 'Rebeus Hagrid',
            profile_image: '/profile_image_hagrid',
        },
    },
];

jest.mock('../../components/button/button', () => ({ Button: (props: any) => props.children }));
jest.mock('../../components/image/image', () => ({ Image: () => 'Image' }));
jest.mock('../../const/const', () => ({ ...jest.requireActual('../../const/const'), noOfImages: 2 }));
jest.mock('../../utils/server/server', () => ({
    server: jest.fn((_, limit, offset) => Promise.resolve(images.slice(offset, offset + limit))),
}));

const flushPromises = async (wrapper: ReactWrapper, extraMethod?: () => void) =>
    act(async () => {
        extraMethod?.();
        await new Promise(setImmediate);
        wrapper.update();
    });

describe('Images', () => {
    it('should match before and after loading images snapshot', async () => {
        const wrapper = mount(<Images columns={2} />);
        expect(wrapper.find({ 'data-test-id': 'grid-item' }).exists()).toBe(false);
        expect(wrapper).toMatchSnapshot();

        await flushPromises(wrapper);
        expect(wrapper.find({ 'data-test-id': 'grid-item' }).length).toBe(4); // twice as many due to mounting of styled components
        expect(wrapper).toMatchSnapshot();
    });

    it('should load more images on load more click', async () => {
        const wrapper = mount(<Images columns={2} />);
        await flushPromises(wrapper);
        expect(wrapper.find({ 'data-test-id': 'grid-item' }).length).toBe(4);

        await flushPromises(wrapper, () => wrapper.find(Button).prop('onClick')());
        expect(wrapper.find({ 'data-test-id': 'grid-item' }).length).toBe(6);
    });
});
