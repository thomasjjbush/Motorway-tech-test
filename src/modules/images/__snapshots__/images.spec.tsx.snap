// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Images should match before and after loading images snapshot 1`] = `
<Images
  desktop={4}
  mobile={1}
  tablet={2}
>
  <styled.div
    columns={2}
    styles="
    display: grid;
    grid-gap: 20px;
    grid-template-columns: repeat(2, 1fr);
    margin: auto;
    max-width: 1300px;
    padding: 20px;
"
  >
    <div
      className="sc-bdfBwQ ABmXt"
    >
      <div
        key="0"
      />
      <div
        key="1"
      />
      <Button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Load more
      </Button>
    </div>
  </styled.div>
</Images>
`;

exports[`Images should match before and after loading images snapshot 2`] = `
<Images
  desktop={4}
  mobile={1}
  tablet={2}
>
  <styled.div
    columns={2}
    styles="
    display: grid;
    grid-gap: 20px;
    grid-template-columns: repeat(2, 1fr);
    margin: auto;
    max-width: 1300px;
    padding: 20px;
"
  >
    <div
      className="sc-bdfBwQ ABmXt"
    >
      <div
        key="0"
      >
        <styled.div
          data-test-id="grid-item"
          key="1"
          styles="
    margin-bottom: 20px;
    position: relative;

    :hover,
    :focus {
        > picture img {
            filter: grayscale(100%);
        }

        > .sc-gsTCUz {
            opacity: 1;
        }
    }
"
          tabIndex={0}
        >
          <div
            className="sc-dlfnbm eNjwaX"
            data-test-id="grid-item"
            tabIndex={0}
          >
            <Image
              alt="alt description"
              src="http://localhost:5000/endpoint"
            >
              Image
            </Image>
            <styled.div
              styles="
    align-items: center;
    background-color: theme.colors.yellow;
    bottom: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    left: 0;
    opacity: 0;
    padding: 20px;
    position: absolute;
    right: 0;
    transition: opacity 0.5s ease;
"
            >
              <div
                className="sc-gsTCUz grUumz"
              >
                <Image
                  alt="Lord Souron"
                  round={true}
                  src="http://localhost:5000/profile_image_souron"
                  width="75px"
                >
                  Image
                </Image>
                <styled.h2
                  styles="
    margin: 0;
"
                >
                  <h2
                    className="sc-hKgILt ioRZeK"
                  >
                    Lord Souron
                  </h2>
                </styled.h2>
                <styled.p
                  styles="
    margin: 0;
"
                >
                  <p
                    className="sc-eCssSg bRCGol"
                  >
                    Likes: 
                    500
                  </p>
                </styled.p>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </div>
      <div
        key="1"
      >
        <styled.div
          data-test-id="grid-item"
          key="2"
          styles="
    margin-bottom: 20px;
    position: relative;

    :hover,
    :focus {
        > picture img {
            filter: grayscale(100%);
        }

        > .sc-gsTCUz {
            opacity: 1;
        }
    }
"
          tabIndex={0}
        >
          <div
            className="sc-dlfnbm eNjwaX"
            data-test-id="grid-item"
            tabIndex={0}
          >
            <Image
              alt="another alt description"
              src="http://localhost:5000/another-endpoint"
            >
              Image
            </Image>
            <styled.div
              styles="
    align-items: center;
    background-color: theme.colors.yellow;
    bottom: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    left: 0;
    opacity: 0;
    padding: 20px;
    position: absolute;
    right: 0;
    transition: opacity 0.5s ease;
"
            >
              <div
                className="sc-gsTCUz grUumz"
              >
                <Image
                  alt="Samwise the brave"
                  round={true}
                  src="http://localhost:5000/profile_image_sam"
                  width="75px"
                >
                  Image
                </Image>
                <styled.h2
                  styles="
    margin: 0;
"
                >
                  <h2
                    className="sc-hKgILt ioRZeK"
                  >
                    Samwise the brave
                  </h2>
                </styled.h2>
                <styled.p
                  styles="
    margin: 0;
"
                >
                  <p
                    className="sc-eCssSg bRCGol"
                  >
                    Likes: 
                    500
                  </p>
                </styled.p>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </div>
      <Button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Load more
      </Button>
    </div>
  </styled.div>
</Images>
`;
